cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -536,6 +536,37 @@ private void writeToHeartBeatTable(@Nullable RunRecordDetail recordedRunRecord,
         // Publish the program STARTING state before starting the program
         programStateWriter.start(programRunId, newProgramOptions, null, programDescriptor);
 
+        long provisioningTime = System.currentTimeMillis() / 1000 -
+          RunIds.getTime(programRunId.getRun(), TimeUnit.SECONDS);
+        // emit provisioning time metric
+        ProfileId profileId = null;
+        try {
+          RunRecordDetail runRecordMeta = appMetadataStore.getRun(programRunId);
+          profileId = SystemArguments
+            .getProfileIdFromArgs(programRunId.getNamespaceId(),
+              runRecordMeta.getSystemArgs()).orElse(null);
+        } catch (IOException e) {
+          LOG.error("Failed to get run record", e);
+        }
+
+        if (profileId != null) {
+          Map<String, String> args = programOptions.getArguments().asMap();
```

Comment:
move this to a private function to have better separation of concern

Reply:
done.